pkg/streamingpromql/aggregations/avg.go

```diff
@@ -0,0 +1,297 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/promql/engine.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors
+
+package aggregations
+
+import (
+	"math"
+
+	"github.com/prometheus/prometheus/model/histogram"
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/floats"
+	"github.com/grafana/mimir/pkg/streamingpromql/functions"
+	"github.com/grafana/mimir/pkg/streamingpromql/limiting"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+type AvgAggregationGroup struct {
+	floats                 []float64
+	floatMeans             []float64
+	floatCompensatingMeans []float64 // Mean, or "compensating value" for Kahan summation.
+	incrementalMeans       []bool    // True after reverting to incremental calculation of the mean value.
+	floatPresent           []bool
+	histograms             []*histogram.FloatHistogram
+	histogramPointCount    int
+
+	// Keeps track of how many series we have encountered thus far for the group at this point
+	// This is necessary to do per point (instead of just counting the groups) as a series may have
+	// stale or non-existent values that are not added towards the count.
+	groupSeriesCounts []float64
+}
+
+func (g *AvgAggregationGroup) AccumulateSeries(data types.InstantVectorSeriesData, steps int, start int64, interval int64, memoryConsumptionTracker *limiting.MemoryConsumptionTracker, emitAnnotationFunc functions.EmitAnnotationFunc) error {
+	var err error
+
+	if g.groupSeriesCounts == nil {
+		g.groupSeriesCounts, err = types.Float64SlicePool.Get(steps, memoryConsumptionTracker)
+		if err != nil {
+			return err
+		}
+		g.groupSeriesCounts = g.groupSeriesCounts[:steps]
+
+	}
+
+	if len(data.Floats) > 0 && g.floats == nil {
+		// First series with float values for this group, populate it.
+		g.floats, err = types.Float64SlicePool.Get(steps, memoryConsumptionTracker)
+		if err != nil {
+			return err
+		}
+
+		g.floatCompensatingMeans, err = types.Float64SlicePool.Get(steps, memoryConsumptionTracker)
+		if err != nil {
+			return err
+		}
+
+		g.floatPresent, err = types.BoolSlicePool.Get(steps, memoryConsumptionTracker)
+		if err != nil {
+			return err
+		}
+
+		g.incrementalMeans, err = types.BoolSlicePool.Get(steps, memoryConsumptionTracker)
+		if err != nil {
+			return err
+		}
```

Comment:
Could we delay allocating this until we switch to incremental calculation mode? ie. only allocate this on line 113 if it hasn't already been populated.

Most of the time, we won't use incremental calculation mode, so this won't be needed, and we can use `g.incrementalMeans == nil` to mean "no samples use incremental calculation mode".